src/TiledArray/algebra/diis.h

```diff
@@ -214,32 +216,84 @@ namespace TiledArray {
 
             // finally, solve the DIIS linear system
             Eigen::ColPivHouseholderQR<EigenMatrixX> A_QR = A.colPivHouseholderQr();
-            c = A_QR.solve(rhs);
+            C_ = A_QR.solve(rhs);
             absdetA = A_QR.absDeterminant();
 
             //std::cout << "DIIS: |A|=" << absdetA << " sol=" << c << std::endl;
 
-            ++nskip;
+            ++nskip_;
 
-          } while (absdetA < zero_determinant && nskip < nvec); // while (system is poorly conditioned)
+          } while (absdetA < zero_determinant && nskip_ < nvec); // while (system is poorly conditioned)
 
           // failed?
           if (absdetA < zero_determinant) {
             std::ostringstream oss;
             oss << "DIIS::extrapolate: poorly-conditioned system, |A| = " << absdetA;
             throw std::domain_error(oss.str());
           }
-          --nskip; // undo the last ++ :-(
+          --nskip_; // undo the last ++ :-(
+
+          {
+
+            zero(x);
+            for (unsigned int k=nskip_, kk=1; k < nvec; ++k, ++kk) {
+              if (not do_mixing || x_extrap_.empty()) {
+                //std::cout << "contrib " << k << " c=" << c[kk] << ":" << std::endl << x_[k] << std::endl;
+                axpy(x, C_[kk], x_[k]);
+                if (extrapolate_error)
+                  axpy(error, C_[kk], errors_[k]);
+              } else {
+                axpy(x, C_[kk] * (1.0 - mixing_fraction), x_[k]);
+                axpy(x, C_[kk] * mixing_fraction, x_extrap_[k]);
+              }
+            }
+          }
+        } // do DIIS
+
+        // only need to keep extrapolated x if doing mixing
+        if (do_mixing) x_extrap_.push_back(x);
+      }
+
+      /// calling this function performs the extrapolation with user-provided
+      /// coefficients.
+      /// \param[in,out] x On input, the most recent solution guess; on output,
+      ///   the extrapolated guess
+      /// \param c user-provided coefficients
+      /// \param nskip number of old vectors to skip
+      void extrapolate(D& x, const EigenVectorX &c, unsigned int nskip = 0) {
```

Comment:
Currently there is quite a bit of code duplication. To avoid code duplication: factor out the computation of the extrapolation parameters (C, nskip) into a separate (public!) function and have the original extrapolate() call compute_extrapolation_parameters() and extrapolate(x,C,nskip) so that there is no duplication of code.